### DISCUSSION TITLE
berry script: how to implement header

### DISCUSSION THREAD

**Discussion by indigmas** [arendst/Tasmota#20764] [Q&A] [answered]
i am planning to steer my e-cars wallbox using berry scripting and "webclient" Class.
i started with an easy example, but no matter how i write the "cl.add_header", i always receive the whole json-response with 6 items.
if i put "https://simple-books-api.glitch.me/books?type=non-fiction" in the browser it unly returns me 2 items
here's the code i tried in berry:
cl = webclient()
cl.begin("https://simple-books-api.glitch.me/books")
cl.add_header('type', 'non-fiction')
r = cl.GET()
print(r)
s = cl.get_string()
print(s)

**Comment by s-hadinger** [comment]
I think you are confusing header and argument. I don't think that any api would change it's response with a header change. 

You must encode yourself the URL if that's what you need

**Reply by sfromis** [reply]
What you have in the browser example is a [query string](https://en.wikipedia.org/wiki/Query_string), after the `?`, not http headers, which are not part of the URL. You need to construct the full query string from Berry, the same way as when you test using a browser.

**Reply by indigmas** [reply]
sorry for confusing, i am trying to make it more clear: if i post the query string without the part after the ? i receive 6 items. if i add "type=non-fiction" i receive only 2 (the ones which are non-fiction). So my question is: how do i manage to get only this 2 "non-fiction" with berry "web-client"?

**Reply by sfromis** [reply, marked as answer]
By doing what I wrote, use the same URL as you're doing in the browser. The `webclient` module has no feature for building query strings. The query string **is** all of what you have after the `?`, there can be no query string part of the URL without that.

**Reply by indigmas** [reply]
that's it :) thanks a lot, it works

**Reply by indigmas** [reply]
can you help me once more? like i wrote in my first post i would like to adapt my wallbox.
below is the format which the wallbox-supplier (easee) defined
is there any way to do it with berry script?
---------
curl --request POST \
     --url https://api.easee.com/api/accounts/login \
     --header 'Authorization: bearer myAuthorizationKey' \
     --header 'accept: application/json' \
     --header 'content-type: application/*+json' \
     --data '
{
  "userName": "myUserName",
  "password": "myPasswort"
}

**Reply by barbudor** [reply]
What are you looking for that you are not already doing in your initial post ?
POST ? => replace `cl.GET()` by `cl.POST()` and use [`json.dump()`](https://github.com/berry-lang/berry/wiki/Chapter-7#dump-function) to convert your map into a JSON string to pass as the payload
Adding headers ? You are already know how to use `cl.add_header()`

**Reply by indigmas** [reply]
this time i need to transfer my userName an pasword and if i just add it like &userName=MyUserName&password=myPasswort it does not work

**Reply by barbudor** [reply]
It's a post request
So you have to place the data as a json string into the payload
Check doc `WebClient. POST()` and json module doc

**Reply by sfromis** [reply]
In addition to formatting the required JSON with userName/password in the payload for the POST, you may also need to use `webclient.add_header` calls to match the 3 `--header` parameters in the `curl` example.

**Reply by sfromis** [reply]
Beware that it looks like need to use the returned access token in the header of requests in the next 24 hours, where you have the string "myAuthorizationKey", but not have this when you make the login request with userName/Password. It also seems possible to refresh the access token without a new login.

https://developer.easee.com/docs/authentication-1
https://developer.easee.com/reference/post_api-accounts-login
https://developer.easee.com/reference/post_api-accounts-refresh-token

**Comment by indigmas** [comment]
sorry, i am still struggling. 
- i placed the data as a json string in the payload
- i also used the 3 add_header
-> i tried to do it this way:
 
cl = webclient()
 cl.begin('https://api.easee.com/api/accounts/login')
 cl.add_header('Authorization', 'bearer myAuthorizationKey')
 cl.add_header('accept', 'application/json')
 cl.add_header('content-type', 'application/*+json')
 r = cl.POST({'userName': 'myusername','password': 'mypassword'})
print(r)
s = cl.get_string()
print(s)

but the error-code is 
stack traceback:
<native>: in native function
<unknown source>: in function 'main'

i even tried 
  r = cl.POST(json.dump({'userName': 'myusername','password': 'mypassword'}))
but i was quite sure from the beginning that that could not work

**Reply by Noschvie** [reply]
cl.add_header('content-type', 'application/*+json')

... seems to be wrong, should be 'application/json' , isn'it?

**Reply by barbudor** [reply]
@Noschvie This should not create a crash anyhow

@indigmas 'json.dump' IS the way to go
May be try to assign the json.dump to an intermediate variable and then pass it as 1st argument to POST

**Reply by barbudor** [reply]
I have no problem executig this code in the console
```
import json
cl = webclient()
cl.begin('https://httpbin.org/post')
cl.add_header("content-type", "application/json")
cl.add_header("accept", "application/json")
payload = json.dump({'userName':'myUserName', 'password':'myPassword'})
r = cl.POST(payload)
print(r)
s = cl.get_string()
print(s)

200
{
  "args": {}, 
  "data": "{\"password\":\"myPassword\",\"userName\":\"myUserName\"}", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "application/json", 
    "Accept-Encoding": "identity;q=1,chunked;q=0.1,*;q=0", 
    "Content-Length": "49", 
    "Content-Type": "application/json", 
    "Host": "httpbin.org", 
    "User-Agent": "TasmotaClient", 
    "X-Amzn-Trace-Id": "Root=1-65e81f89-2427fa36561732673745a7cc"
  }, 
  "json": {
    "password": "myPassword", 
    "userName": "myUserName"
  }, 
  "origin": "82.65.83.41", 
  "url": "https://httpbin.org/post"
}

```

**Reply by barbudor** [reply]
And it works the same with the `json.dump` inside the POST call

Did you forgot to `import json`  ?

**Reply by sfromis** [reply]
The reason for the first failure is that the `POST` method expect the payload argument to be a string, not a map; hence using `json.dump` is a good way of converting the map to a JSON-valid string format. Of course, you *could* also build the string yourself, but getting the JSON syntax "just right" can be a hassle when having to manage quotes and variables.

To be clear, while `{'userName': 'myusername','password': 'mypassword'}` almost looks like it is JSON, it is actually not, as JSON uses `"` as string delimiter. It is a Berry way of creating a map without doing it field by field, and that's why it has to be converted to a string. While Berry also supports `"` as string delimiter, it would still not be JSON, as the whole would not be a string.

**Reply by indigmas** [reply]
that was the issue, i really did not put the `import json` 
so view seconds ago, for the very first time i told my wallbox to change charger current by berry script :) :)
thank you very much for your support and vor your patienc.
But as i am a beginner this was neccessary ;)

**Reply by Noschvie** [reply]
Great for you and for Berry too ! Have fun with your devices :-)
